src/scene/materials/node-material.js

```diff
@@ -0,0 +1,692 @@
+                        }
+
+                    } else {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+        return false;
+    },
+
+    // NB only called when shader graph is valid and shader is marked as dirty
+    initShader: function (device) {
+        // this is where we could get a list of pass types in current app render pipeline (layers)
+        // and query shader graph to check if there is a special output
+        // var passes=[SHADER_DEPTH, SHADER_FORWARD, SHADER_FORWARDHDR, SHADER_PICK, SHADER_SHADOW];
+        var passes = [SHADER_FORWARD];
+
+        for (var i = 0; i < passes.length; i++) {
+            var useSkin = false;
+
+            if (this.meshInstances && this.meshInstances[0] && this.meshInstances[0].skinInstance) {
+                useSkin = true; // TODO make this a variant?
+            }
+
+            var options = {
+                skin: useSkin,
+                shaderGraph: this,
+                pass: passes[i],
+                maxPixelLights: (this.maxPixelLights) ? this.maxPixelLights : 4,
+                maxVertexLights: (this.maxVertexLights) ? this.maxVertexLights : 8
+            };
+
+            // var depGraphTextureList=[];
+
+            var shaderDefinition = programlib.node.createShaderDefinition(device, options);
+            this.shaderVariants[passes[i]] = new Shader(device, shaderDefinition);
+        }
+    },
+
+    _addGraphVar: function (type, name, value) {
+        var graphVar;
+        if (value instanceof Texture) {
+            graphVar = { type: type, name: name, valueTex: value };
+        } else if (value instanceof Vec4) {
+            graphVar = { type: type, name: name, valueX: value.x, valueY: value.y, valueZ: value.z, valueW: value.w };
+        } else if (value instanceof Vec3) {
+            graphVar = { type: type, name: name, valueX: value.x, valueY: value.y, valueZ: value.z };
+        } else if (value instanceof Vec2) {
+            graphVar = { type: type, name: name, valueX: value.x, valueY: value.y };
+        } else if (typeof(value) === 'number') {
+            graphVar = { type: type, name: name, valueX: value };
+        } else {
+            graphVar = { type: type, name: name };
+        }
+
+        this.graphData.graphVars.push(graphVar);
+
+        return graphVar;
+    },
+
+    addInput: function (type, name, value) {
+        return this._addGraphVar(type, 'IN_' + name, value);
+    },
+
+    addOutput: function (type, name, value) {
+        return this._addGraphVar(type, 'OUT_' + name, value);
+    },
+
+    addConstant: function (type, value) {
+        return this._addGraphVar(type, 'CONST_' + type + '_' + this.graphData.graphVars.length, value); // create a unique name
+    },
+
+    genCustomFuncVars: function () {
+        var functionString = this.graphData.customFuncGlsl.trim();
+
+        var head = functionString.split(")")[0];

```

Comment:
It's currently quite restrictive - for the very reason you mention: that making a robust GLSL parser would be a big task! The parsed variables meta data are actually stored in the node material asset - along with the user defined values - so the parsing is should only be done when the editor event fires to say that the GLSL shader asset is updated. As a long term goal, I would love to be able to take any GLSL code and make it work with shader graphs - imagine just copying and pasting from ShaderToy? Definitely something to aim for. For now, good documentation and examples (when released publicly) - like what we have for script attributes (which also have meet PlayCanvas specific semantics) could be a good start? Another thing is that, as mentioned in the last graphics sync, we will probably have to add a new text asset type that is only viewable and usable in the visual shader editor - this may allow us to add some UX/Autocomplete functionality - which can help users make conformant GLSL functions?

Reply:
I will bear this in mind - and let's revisit in a later PR?